File: app/src/main/java/com/digicoffer/lauditor/TimeSheets/ViewModels/SubmittedTimeSheets.java
package com.digicoffer.lauditor.TimeSheets.ViewModels;

import android.app.Dialog;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import com.digicoffer.lauditor.R;
import com.digicoffer.lauditor.TimeSheets.Adapters.TimeSheetsAdapter;
import com.digicoffer.lauditor.TimeSheets.Models.EventsModel;
import com.digicoffer.lauditor.TimeSheets.Models.StatusModel;
import com.digicoffer.lauditor.TimeSheets.Models.TSMatterModel;
import com.digicoffer.lauditor.TimeSheets.Models.TasksModel;
import com.digicoffer.lauditor.TimeSheets.Models.TimeSheetModel;
import com.digicoffer.lauditor.TimeSheets.Models.WeekModel;
import com.digicoffer.lauditor.TimeSheets.Models.WeekTotalModel;
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener;
import com.digicoffer.lauditor.Webservice.HttpResultDo;
import com.digicoffer.lauditor.Webservice.CommonApiHelper.WebServiceHelper;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;

public class SubmittedTimeSheets extends Fragment implements AsyncTaskCompleteListener {
    private View view;
    String date = "";
    boolean issubmitted = true;
    String current_date = "";
    private CardView cv_details_activity_log;
    private Dialog progressDialog;
    private ArrayList<TimeSheetModel> timeSheetsList = new ArrayList<>();
    private ArrayList<TSMatterModel> matterList = new ArrayList<>();
    private ArrayList<TasksModel> tasksList = new ArrayList<>();
    private ArrayList<EventsModel> eventsList = new ArrayList<>();
    private ArrayList<StatusModel> statusList = new ArrayList<>();
    private ArrayList<WeekTotalModel> weektotalList = new ArrayList<>();
    private ArrayList<WeekModel> weeksList = new ArrayList<>();
    RecyclerView rv_submitted_timesheets;

    // Empty state views
    LinearLayout layout_empty_state;
    TextView tv_empty_subtitle,tv_title;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.submitted_timesheets, container, false);
        Bundle bundle = getArguments();
        date = bundle.getString("date");
        ArrayList<String> weekDates = bundle.getStringArrayList("weekDates");
        current_date = weekDates.get(0);
        rv_submitted_timesheets = view.findViewById(R.id.rv_submitted_timesheets);

        // Bind empty state views
        layout_empty_state = view.findViewById(R.id.layout_empty_state);
        tv_empty_subtitle = view.findViewById(R.id.tv_empty_subtitle);
        tv_title = view.findViewById(R.id.tv_title);

        if (date.isEmpty()) {
            callCurrentDateTimeSheetsWebservice(date);
        } else if (date == null) {
            callCurrentDateTimeSheetsWebservice(date);
        } else {
            callTimeSheetsWebservice(date);
        }

        SimpleDateFormat inputFormat = new SimpleDateFormat("dd-MM-yyyy", Locale.ENGLISH);
        SimpleDateFormat outputFormat = new SimpleDateFormat("EEE MMM d,yyyy", Locale.ENGLISH);

        ArrayList<String> formattedDates = new ArrayList<>();

        for (String dateStr : weekDates) {
            try {
                Date date1 = inputFormat.parse(dateStr);
                String formattedDate = outputFormat.format(date1);
                formattedDates.add(formattedDate);
            } catch (ParseException e) {
                e.fillInStackTrace();
            }
        }
        for (String value : formattedDates) {
            weeksList.add(new WeekModel(value));
        }

        return view;
    }

    private void callCurrentDateTimeSheetsWebservice(String date) {
        try {
            clearList();
            progressDialog = AndroidUtils.get_progress(getActivity());
            JSONObject data = new JSONObject();
            WebServiceHelper.callHttpWebService(this, getContext(), WebServiceHelper.RestMethodType.GET, "v3/user/timesheets" + "?submitted=true", "TimeSheets", data.toString());
        } catch (Exception e) {
            e.fillInStackTrace();
        }
    }

    private void callTimeSheetsWebservice(String date) {
        try {
            clearList();
            progressDialog = AndroidUtils.get_progress(getActivity());
            JSONObject data = new JSONObject();
            SimpleDateFormat inputFormat = new SimpleDateFormat("dd-MM-yyyy", Locale.US);
            SimpleDateFormat outputFormat = new SimpleDateFormat("dd-MM-yyyy", Locale.US);
            Date new_date = inputFormat.parse(date);
            String outputDate = outputFormat.format(new_date);
            WebServiceHelper.callHttpWebService(this, getContext(), WebServiceHelper.RestMethodType.GET, "v3/user/timesheets/" + outputDate + "?submitted=true", "TimeSheets", data.toString());
        } catch (Exception e) {
            e.fillInStackTrace();
        }
    }

    private void clearList() {
        timeSheetsList.clear();
        weektotalList.clear();
        eventsList.clear();
        matterList.clear();
        statusList.clear();
        tasksList.clear();
    }

    @Override
    public void onClick(View view) {

    }

    @Override
    public void onAsyncTaskComplete(HttpResultDo httpResult) {
        if (progressDialog.isShowing() && progressDialog != null) {
            AndroidUtils.dismiss_dialog(progressDialog);
        }
        if (httpResult.getResult() == WebServiceHelper.ServiceCallStatus.Success) {
            try {
                JSONObject result = new JSONObject(httpResult.getResponseContent());
                if (httpResult.getRequestType().equals("TimeSheets")) {
                    JSONObject dates = result.getJSONObject("dates");
                    timeSheetsList.clear();
                    loadTimesheetData(dates, result);
                }
            } catch (Exception e) {
                e.fillInStackTrace();
            }
        } else if (httpResult.getResult() == WebServiceHelper.ServiceCallStatus.Failed) {
            if (progressDialog != null && progressDialog.isShowing())
                AndroidUtils.dismiss_dialog(progressDialog);
            try {
                JSONObject result = new JSONObject(httpResult.getResponseContent());
                AndroidUtils.showErrorAlert(result.optString("msg"), getActivity());
            } catch (Exception e) {
                e.fillInStackTrace();
            }
        } else {
            if (progressDialog != null && progressDialog.isShowing())
                AndroidUtils.dismiss_dialog(progressDialog);
            AndroidUtils.showErrorAlert(httpResult.getResponseContent().toString(), getActivity());
        }
    }

    private void loadTimesheetData(JSONObject dates, JSONObject result) throws JSONException {
        timeSheetsList.clear();
        TimeSheetModel timeSheetModel = new TimeSheetModel();
        timeSheetModel.setFrozen(dates.getBoolean("isFrozen"));
        timeSheetModel.setCurrentWeek(dates.getString("currentWeek"));
        timeSheetModel.setNextWeek(dates.getString("nextWeek"));
        timeSheetModel.setPrevWeek(dates.getString("prevWeek"));
        JSONObject timesheets = result.getJSONObject("timesheetList");

        JSONArray matters = timesheets.getJSONArray("matters");
        JSONObject weektotal = timesheets.getJSONObject("weekTotal");
        WeekTotalModel weekModel = new WeekTotalModel();
        weekModel.setMon(weektotal.getString("Mon"));
        weekModel.setTue(weektotal.getString("Tue"));
        weekModel.setWed(weektotal.getString("Wed"));
        weekModel.setThu(weektotal.getString("Thu"));
        weekModel.setFri(weektotal.getString("Fri"));
        weekModel.setSat(weektotal.getString("Sat"));
        weekModel.setSun(weektotal.getString("Sun"));
        weektotalList.add(weekModel);

        for (int i = 0; i < matters.length(); i++) {
            JSONObject jsonObject = matters.getJSONObject(i);
            TSMatterModel matterModel = new TSMatterModel();
            matterModel.setMattername(jsonObject.getString("matterName"));
            matterModel.setMatterid(jsonObject.getString("matterId"));
            matterModel.setTasks(jsonObject.getJSONArray("tasks"));
            if (jsonObject.has("matterType")) {
                matterModel.setMatter_type(jsonObject.getString("matterType"));
            }
            matterList.add(matterModel);
        }

        timeSheetsList.add(timeSheetModel);

        for (int j = 0; j < matterList.size(); j++) {
            for (int m = 0; m < matterList.get(j).getTasks().length(); m++) {
                JSONObject jsonObject = matterList.get(j).getTasks().getJSONObject(m);
                EventsModel eventsModel = new EventsModel();
                if (jsonObject.has("billing")) {
                    eventsModel.setBilling(jsonObject.getString("billing"));
                }
                eventsModel.setTaskName(jsonObject.getString("taskName"));
                eventsModel.setTotal(jsonObject.getString("total"));
                eventsModel.setMon(jsonObject.getJSONObject("Mon"));
                eventsModel.setTue(jsonObject.getJSONObject("Tue"));
                eventsModel.setWed(jsonObject.getJSONObject("Wed"));
                eventsModel.setThu(jsonObject.getJSONObject("Thu"));
                eventsModel.setFri(jsonObject.getJSONObject("Fri"));
                eventsModel.setSat(jsonObject.getJSONObject("Sat"));
                eventsModel.setSun(jsonObject.getJSONObject("Sun"));
                eventsModel.setMatter_id(matterList.get(j).getMatterid());
                eventsModel.setMatter_name(matterList.get(j).getMattername());
                eventsList.add(eventsModel);
            }
        }

        if (timeSheetModel.isFrozen()) {
            updateEmptyState();
            loadTimesheetsRecyclerview();
        } else {
//            tv_title.setText("Timesheet Not Submitted PLease Select Other Week");
//            tv_empty_subtitle.setVisibility(View.GONE);
            // timesheet is not frozen — treat as empty / not submitted
            showEmptyState("Timesheets have not been submitted yet.");
        }
    }

    private void updateEmptyState() {
        if (eventsList.isEmpty()) {
            layout_empty_state.setVisibility(View.VISIBLE);
            rv_submitted_timesheets.setVisibility(View.GONE);
            tv_title.setText("Timesheet Not Submitted PLease Select Other Week");
            tv_empty_subtitle.setVisibility(View.GONE);
        } else {
            layout_empty_state.setVisibility(View.GONE);
            rv_submitted_timesheets.setVisibility(View.VISIBLE);
        }
    }

    private void showEmptyState(String message) {
        layout_empty_state.setVisibility(View.VISIBLE);
        rv_submitted_timesheets.setVisibility(View.GONE);
        tv_empty_subtitle.setText(message);
    }

    private void loadTimesheetsRecyclerview() {
        TimeSheetsAdapter timeSheetsAdapter = new TimeSheetsAdapter(weeksList, eventsList, weektotalList, getContext(), issubmitted, null, getActivity());
        rv_submitted_timesheets.setAdapter(timeSheetsAdapter);
        AndroidUtils.LoadingRecyclerview(rv_submitted_timesheets, getContext());
        AndroidUtils.setupBottomSpacerFooter(rv_submitted_timesheets, getResources().getDimensionPixelSize(R.dimen.twentyeight_dp));
        if (timeSheetsAdapter != null && timeSheetsAdapter.getItemCount() > 0) {
            int lastPosition = timeSheetsAdapter.getItemCount() - 1;
            rv_submitted_timesheets.smoothScrollToPosition(lastPosition);
        }
    }
}

File: app/src/main/java/com/digicoffer/lauditor/TimeSheets/ViewModels/SubmittedTimeSheets.kt
package com.digicoffer.lauditor.TimeSheets.ViewModels

import android.app.Dialog
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import com.digicoffer.lauditor.R
import com.digicoffer.lauditor.TimeSheets.Adapters.TimeSheetsAdapter
import com.digicoffer.lauditor.TimeSheets.Adapters.WeeklyTSAdapter
import com.digicoffer.lauditor.TimeSheets.Models.EventsModel
import com.digicoffer.lauditor.TimeSheets.Models.StatusModel
import com.digicoffer.lauditor.TimeSheets.Models.TSMatterModel
import com.digicoffer.lauditor.TimeSheets.Models.TaskModel
import com.digicoffer.lauditor.TimeSheets.Models.TasksModel
import com.digicoffer.lauditor.TimeSheets.Models.TimeSheetModel
import com.digicoffer.lauditor.TimeSheets.Models.WeekModel
import com.digicoffer.lauditor.TimeSheets.Models.WeekTotalModel
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener
import com.digicoffer.lauditor.Webservice.HttpResultDo
import com.digicoffer.lauditor.Webservice.CommonApiHelper.WebServiceHelper
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Date
import java.util.Locale

class SubmittedTimeSheets : Fragment(), AsyncTaskCompleteListener, View.OnClickListener {
    private var view: View? = null
    var date = ""
    var issubmitted = true
    var current_date = ""
    private var cv_details_activity_log: CardView? = null
    private var progressDialog: Dialog? = null
    private val timeSheetsList = ArrayList<TimeSheetModel>()
    private val matterList = ArrayList<TSMatterModel>()
    private val tasksList = ArrayList<TasksModel>()
    private val eventsList = ArrayList<EventsModel>()
    private val statusList = ArrayList<StatusModel>()
    private val weektotalList = ArrayList<WeekTotalModel>()
    private val weeksList = ArrayList<WeekModel>()
    var rv_submitted_timesheets: RecyclerView? = null

    var layout_empty_state: LinearLayout? = null
    var tv_empty_subtitle: TextView? = null
    var tv_title: TextView? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.submitted_timesheets, container, false)
        this.view = view
        val bundle = arguments
        if (bundle != null) {
            date = bundle.getString("date") ?: ""
            val weekDates = bundle.getStringArrayList("weekDates")
            if (weekDates != null && weekDates.isNotEmpty()) {
                current_date = weekDates[0]

                val inputFormat = SimpleDateFormat("dd-MM-yyyy", Locale.ENGLISH)
                val outputFormat = SimpleDateFormat("EEE MMM d,yyyy", Locale.ENGLISH)
                val formattedDates = ArrayList<String>()

                for (dateStr in weekDates) {
                    try {
                        val d1 = inputFormat.parse(dateStr)
                        if (d1 != null) {
                            formattedDates.add(outputFormat.format(d1))
                        }
                    } catch (e: ParseException) {
                        e.fillInStackTrace()
                    }
                }
                for (value in formattedDates) {
                    weeksList.add(WeekModel(value))
                }
            }
        }

        rv_submitted_timesheets = view.findViewById(R.id.rv_submitted_timesheets)
        layout_empty_state = view.findViewById(R.id.layout_empty_state)
        tv_empty_subtitle = view.findViewById(R.id.tv_empty_subtitle)
        tv_title = view.findViewById(R.id.tv_title)

        if (date.isEmpty()) {
            callCurrentDateTimeSheetsWebservice(date)
        } else {
            callTimeSheetsWebservice(date)
        }

        return view
    }

    private fun callCurrentDateTimeSheetsWebservice(date: String?) {
        try {
            clearList()
            progressDialog = AndroidUtils.get_progress(requireActivity())
            val data = JSONObject()
            WebServiceHelper.callHttpWebService(
                this,
                requireContext(),
                WebServiceHelper.RestMethodType.GET,
                "v3/user/timesheets?submitted=true",
                "TimeSheets",
                data.toString()
            )
        } catch (e: Exception) {
            e.fillInStackTrace()
        }
    }

    private fun callTimeSheetsWebservice(date: String?) {
        try {
            clearList()
            progressDialog = AndroidUtils.get_progress(requireActivity())
            val data = JSONObject()
            val inputFormat = SimpleDateFormat("dd-MM-yyyy", Locale.US)
            val outputFormat = SimpleDateFormat("dd-MM-yyyy", Locale.US)
            val new_date = inputFormat.parse(date!!)
            val outputDate = outputFormat.format(new_date!!)
            WebServiceHelper.callHttpWebService(
                this,
                requireContext(),
                WebServiceHelper.RestMethodType.GET,
                "v3/user/timesheets/$outputDate?submitted=true",
                "TimeSheets",
                data.toString()
            )
        } catch (e: Exception) {
            e.fillInStackTrace()
        }
    }

    private fun clearList() {
        timeSheetsList.clear()
        weektotalList.clear()
        eventsList.clear()
        matterList.clear()
        statusList.clear()
        tasksList.clear()
    }

    override fun onClick(view: View) {}

    override fun onAsyncTaskComplete(httpResult: HttpResultDo) {
        if (progressDialog != null && progressDialog!!.isShowing) {
            AndroidUtils.dismiss_dialog(progressDialog)
        }
        if (httpResult.result == WebServiceHelper.ServiceCallStatus.Success) {
            try {
                val result = JSONObject(httpResult.responseContent)
                if (httpResult.requestType == "TimeSheets") {
                    val dates = result.getJSONObject("dates")
                    timeSheetsList.clear()
                    loadTimesheetData(dates, result)
                }
            } catch (e: Exception) {
                e.fillInStackTrace()
            }
        } else if (httpResult.result == WebServiceHelper.ServiceCallStatus.Failed) {
            if (progressDialog != null && progressDialog!!.isShowing) {
                AndroidUtils.dismiss_dialog(progressDialog)
            }
            try {
                val result = JSONObject(httpResult.responseContent)
                AndroidUtils.showErrorAlert(result.optString("msg"), activity)
            } catch (e: Exception) {
                e.fillInStackTrace()
            }
        } else {
            if (progressDialog != null && progressDialog!!.isShowing) {
                AndroidUtils.dismiss_dialog(progressDialog)
            }
            AndroidUtils.showErrorAlert(httpResult.responseContent.toString(), activity)
        }
    }

    private fun loadTimesheetData(dates: JSONObject, result: JSONObject) {
        timeSheetsList.clear()
        val timeSheetModel = TimeSheetModel()
        timeSheetModel.isFrozen = dates.getBoolean("isFrozen")
        timeSheetModel.currentWeek = dates.getString("currentWeek")
        timeSheetModel.nextWeek = dates.getString("nextWeek")
        timeSheetModel.prevWeek = dates.getString("prevWeek")
        val timesheets = result.getJSONObject("timesheetList")

        val matters = timesheets.getJSONArray("matters")
        val weektotal = timesheets.getJSONObject("weekTotal")
        val weekModel = WeekTotalModel().apply {
            Mon = weektotal.getString("Mon")
            Tue = weektotal.getString("Tue")
            Wed = weektotal.getString("Wed")
            Thu = weektotal.getString("Thu")
            Fri = weektotal.getString("Fri")
            Sat = weektotal.getString("Sat")
            Sun = weektotal.getString("Sun")
        }
        weektotalList.add(weekModel)

        for (i in 0 until matters.length()) {
            val jsonObject = matters.getJSONObject(i)
            val matterModel = TSMatterModel().apply {
                mattername = jsonObject.getString("matterName")
                matterid = jsonObject.getString("matterId")
                Tasks = jsonObject.getJSONArray("tasks")
                if (jsonObject.has("matterType")) {
                    matter_type = jsonObject.getString("matterType")
                }
            }
            matterList.add(matterModel)
        }

        timeSheetsList.add(timeSheetModel)

        for (j in 0 until matterList.size) {
            val tasks = matterList[j].Tasks
            if (tasks != null) {
                for (m in 0 until tasks.length()) {
                    val jsonObject = tasks.getJSONObject(m)
                    val eventsModel = EventsModel()
                    if (jsonObject.has("billing")) {
                        eventsModel.billing = jsonObject.getString("billing")
                    }
                    eventsModel.taskName = jsonObject.getString("taskName")
                    eventsModel.total = jsonObject.getString("total")
                    eventsModel.mon = jsonObject.getJSONObject("Mon")
                    eventsModel.tue = jsonObject.getJSONObject("Tue")
                    eventsModel.wed = jsonObject.getJSONObject("Wed")
                    eventsModel.thu = jsonObject.getJSONObject("Thu")
                    eventsModel.fri = jsonObject.getJSONObject("Fri")
                    eventsModel.sat = jsonObject.getJSONObject("Sat")
                    eventsModel.sun = jsonObject.getJSONObject("Sun")
                    eventsModel.matter_id = matterList[j].matterid
                    eventsModel.matter_name = matterList[j].mattername
                    eventsList.add(eventsModel)
                }
            }
        }

        if (timeSheetModel.isFrozen) {
            updateEmptyState()
            loadTimesheetsRecyclerview()
        } else {
            showEmptyState("Timesheets have not been submitted yet.")
        }
    }

    private fun updateEmptyState() {
        if (eventsList.isEmpty()) {
            layout_empty_state?.visibility = View.VISIBLE
            rv_submitted_timesheets?.visibility = View.GONE
            tv_title?.text = "Timesheet Not Submitted PLease Select Other Week"
            tv_empty_subtitle?.visibility = View.GONE
        } else {
            layout_empty_state?.visibility = View.GONE
            rv_submitted_timesheets?.visibility = View.VISIBLE
        }
    }

    private fun showEmptyState(message: String) {
        layout_empty_state?.visibility = View.VISIBLE
        rv_submitted_timesheets?.visibility = View.GONE
        tv_empty_subtitle?.text = message
    }

    private fun loadTimesheetsRecyclerview() {
        val timeSheetsAdapter = TimeSheetsAdapter(
            weeksList,
            eventsList,
            weektotalList,
            requireContext(),
            issubmitted,
            object : WeeklyTSAdapter.InterfaceListener {
                override fun DeleteEvent_Timesheet(taskModel: TaskModel) {}
                override fun EditEvent_Timesheet(eventsModels: TaskModel, date: String, matter_type: String) {}
            },
            requireActivity()
        )
        rv_submitted_timesheets?.adapter = timeSheetsAdapter
        AndroidUtils.LoadingRecyclerview(rv_submitted_timesheets, context)
        AndroidUtils.setupBottomSpacerFooter(
            rv_submitted_timesheets,
            resources.getDimensionPixelSize(R.dimen.twentyeight_dp)
        )
        if (timeSheetsAdapter.itemCount > 0) {
            val lastPosition = timeSheetsAdapter.itemCount - 1
            rv_submitted_timesheets?.smoothScrollToPosition(lastPosition)
        }
    }
}
